### PR TITLE
nixos/restic: add 'backups.package' option to override the restic package

### LINKED ISSUE #176314: expose services.restic.package option for restic backup module
### Describe the bug
I'd like to be able to customize the package in the `restic` backup service with some shell script wrappers. Can we get a `services.restic.package` option to compliment the current `services.restic.server.package`?

### Notify maintainers
@mbrgm

### PR DESCRIPTION
**Pull request by alan-strohm**
Fixes #176314

Also allows for using a wrapped version of restic with special
capabilities as described in the [restic docs](https://restic.readthedocs.io/en/latest/080_examples.html#backing-up-your-system-without-running-restic-as-root)

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by winterqt** [comment]
@ofborg test restic

**Review by teutat3s — APPROVED** [review]

File: nixos/modules/services/backup/restic.nix [hunk 1]
```diff
@@ -222,6 +222,13 @@ in
             A script that must run after finishing the backup process.
           '';
         };
+
+        package = mkOption {
+          default = pkgs.restic;
+          defaultText = literalExpression "pkgs.restic";
+          type = types.package;
+          description = "Restic package to use.";
```
Comment (winterqt) [inline nixos/modules/services/backup/restic.nix:230]:
```suggestion
          type = types.package;
          default = pkgs.restic;
          defaultText = literalExpression "pkgs.restic";
          description = ''
            Restic package to use.
          '';
```

**Comment by winterqt** [comment]
Just to sanity check...

@ofborg test restic

**Comment by winterqt** [comment]
This implementation looks good.

However, I'm curious: what would this allow (other than privilege dropping, which we may be able to come up with a better solution for) that the pre-backup and post-backup hooks don't? I've also asked this in the issue, but posting here for posterity.

File: nixos/tests/restic.nix [hunk 2]
```diff
@@ -63,6 +63,12 @@ import ./make-test-python.nix (
               inherit repository passwordFile;
               pruneOpts = [ "--keep-last 1" ];
             };
+            custompackage = {
+              inherit repository passwordFile paths;
+              package = pkgs.writeShellScriptBin "restic" ''
+                echo "$@" >> /tmp/fake-restic.log;
+                '';
```
Comment (winterqt) [inline nixos/tests/restic.nix:70]:
```suggestion
              '';
```

**Review by winterqt — APPROVED** [review]

**Pull request merged**